match2/commons/match_group_display_bracket.lua

```diff
@@ -1,738 +1,745 @@
+local String = require('Module:String')
 local Table = require('Module:Table')
-local WikiSpecific = require('Module:Brkts/WikiSpecific')
+local TypeUtil = require('Module:TypeUtil')
 
-local getArgs = require('Module:Arguments').getArgs
-local utils = require('Module:LuaUtils')
 local html = mw.html
 
-local ZERO_WIDTH_SPACE = '&#8203;'
-local COMPENSATABLE_SINGLES = 2
-local NIL_EXTRADATA = {0, 0, 0, 0}
-local INFINITE_HEIGHT = 10000
+local Bracket = {propTypes = {}, types = {}}
+
+-- Called by MatchGroup/Display
+function Bracket.luaGet(_, args)
+    return Bracket.BracketContainer({
+        bracketId = args[1],
+        config = Bracket.configFromArgs(args),
+    })
+end
+
+function Bracket.configFromArgs(args)
+    return {
+        headerHeight = tonumber(args.headerHeight),
+        headerMargin = tonumber(args.headerMargin),
+        hideRoundTitles = LuaUtils.misc.readBool(args.hideRoundTitles),
+        lineWidth = tonumber(args.lineWidth),
+        matchMargin = tonumber(args.matchMargin),
+        matchWidth = tonumber(args.matchWidth),
+        matchWidthMobile = tonumber(args.matchWidthMobile),
+        opponentHeight = tonumber(args.opponentHeight),
+        qualifiedHeader = args.qualifiedHeader,
+        roundHorizontalMargin = tonumber(args.roundHorizontalMargin),
+        scoreWidth = tonumber(args.scoreWidth),
+    }
+end
 
--- allowed values for configuring the bracket and their default values
-local BRACKET_CONFIG_FORMAT = {
-    emptyRoundTitles = 'boolean',
+Bracket.types.BracketConfig = TypeUtil.struct({
+    MatchSummaryContainer = 'function',
+    OpponentEntry = 'function',
     headerHeight = 'number',
-    hideMatchLine = 'boolean',
+    headerMargin = 'number',
     hideRoundTitles = 'boolean',
-    matchHeight = 'number',
+    lineWidth = 'number',
+    matchHasDetails = 'function',
+    matchMargin = 'number',
     matchWidth = 'number',
     matchWidthMobile = 'number',
-    qualifiedHeader = 'string',
+    opponentHeight = 'number',
+    qualifiedHeader = 'string?',
+    roundHorizontalMargin = 'number',
     scoreWidth = 'number',
-}
+})
+Bracket.types.BracketConfigOptions = TypeUtil.struct(
+    Table.mapValues(Bracket.types.BracketConfig.struct, TypeUtil.optional)
+)
 
--- default bracket config
-local BRACKET_CONFIG_DEFAULT = {
-    emptyRoundTitles = false,
-    headerHeight = 25,
-    hideMatchLine = false,
-    hideRoundTitles = false,
-    matchHeight = 44,
-    matchWidth = 150,
-    matchWidthMobile = 90,
-    qualifiedHeader = nil,
-    scoreWidth = 20,
+--[[
+Display component for a tournament bracket. The bracket is specified by ID. 
+The component fetches the match data from LPDB or page variables.
+]]
+Bracket.propTypes.BracketContainer = {
+    bracketId = 'string',
+    config = TypeUtil.optional(Bracket.types.BracketConfigOptions),
 }
+function Bracket.BracketContainer(props)
+    DisplayUtil.assertPropTypes(props, Bracket.propTypes.BracketContainer)
```

Comment:
Maybe some other time. I want to finish ffa modules first, and don't want to spend too much time on this.

The type check functionality in TypeUtil was meant as a stop gap, since the components will likely be rewritten in JavaScript eventually. Alternatively someone can set up a Lua equivalent of TypeScript. (what it would be I don't know)

Reply:
> since the components will likely be rewritten in JavaScript eventually

What are you basing this on?

> Maybe some other time. I want to finish ffa modules first, and don't want to spend too much time on the type checker.

I think the documentation is important.

Reply:
>What are you basing this on?

Just a guess based on the features Liquipedia wants to support in the future.

>I think the documentation is important.

Ok, I can write documentation for the type checker sooner then. There should be lots of good opportunities to write documentation in the future.

Reply:
> > What are you basing this on?
> 
> Just a guess based on the features Liquipedia wants to support in the future.

We've mentioned before that this is extremely unlikely. So no point delaying work because of it